lib/spack/llnl/util/filesystem.py

```diff
@@ -236,25 +238,112 @@ def unset_executable_mode(path):
     os.chmod(path, mode)
 
 
-def install(src, dest):
-    """Manually install a file to a particular location."""
-    tty.debug("Installing %s to %s" % (src, dest))
+def copy(src, dst):
+    """Copies the file *src* to the file or directory *dst*.
 
-    # Expand dest to its eventual full path if it is a directory.
-    if os.path.isdir(dest):
-        dest = join_path(dest, os.path.basename(src))
+    If *dst* specifies a directory, the file will be copied into *dst*
+    using the base filename from *src*.
+
+    Parameters:
+        src (str): the file to copy
+        dst (str): the destination file or directory
+    """
+    tty.debug('Copying {0} to {1}'.format(src, dst))
+
+    shutil.copy(src, dst)
+
+
+def install(src, dst):
+    """Installs the file *src* to the file or directory *dst*.
+
+    If *dst* specifies a directory, the file will be installed into *dst*
+    using the base filename from *src*.
+
+    Works similarly to :py:func:`copy` with the addition of setting proper
+    permissions on the installed file.
+
+    Parameters:
+        src (str): the file to install
+        dst (str): the destination file or directory
+    """
+    tty.debug('Installing {0} to {1}'.format(src, dst))
+
+    # Expand dst to its eventual full path if it is a directory.
+    if os.path.isdir(dst):
+        dst = join_path(dst, os.path.basename(src))
+
+    shutil.copy(src, dst)
+
+    set_install_permissions(dst)
+    copy_mode(src, dst)
 
-    shutil.copy(src, dest)
-    set_install_permissions(dest)
-    copy_mode(src, dest)
 
+def copy_tree(src, dst, symlinks=True):
+    """Recursively copy an entire directory tree rooted at *src*.
 
-def install_tree(src, dest, **kwargs):
-    """Manually install a directory tree to a particular location."""
-    tty.debug("Installing %s to %s" % (src, dest))
-    shutil.copytree(src, dest, **kwargs)
+    If the destination directory *dst* does not already exist, it will
+    be created as well as missing parent directories.
+
+    If *symlinks* is true, symbolic links in the source tree are represented
+    as symbolic links in the new tree and the metadata of the original links
+    will be copied as far as the platform allows; if false, the contents and
+    metadata of the linked files are copied to the new tree.
+
+    Parameters:
+        src (str): the directory to copy
+        dst (str): the destination directory
+        symlinks (bool): whether or not to preserve symlinks
+    """
+    tty.debug('Copying {0} to {1}'.format(src, dst))
+
+    mkdirp(dst)
+
+    for s, d in traverse_tree(src, dst, order='pre', follow_nonexisting=True):
+        if symlinks and os.path.islink(s):
+            # Note that this won't rewrite absolute links into the old
+            # root to point at the new root. Should we handle that case?
```

Comment:
Per @healther's earlier github comment (and this comment I made in the code) -- should we change this now to handle absolute paths better?

The problem this is referring to is this.  Say you `install_tree('/path/to/source', '/path/to/dest')`, and `/path/to/source/` looks like this:

```
/path/to/source/
    foo/
        external_abs_link -> /some/other/path/some_file      # this we can't really do much about
        internal_abs_link -> /path/to/source/bar/some_file   # this will be broken after install
        relative_link -> ../bar/some_file                    # this will remain valid
    bar/
        some_file
```

The result right now is:

```
/path/to/dest/
    foo/
        external_abs_link -> /some/other/path/some_file      # keep the same
        internal_abs_link -> /path/to/source/bar/some_file   # probably should be /path/to/dest/bar/some_file
        relative_link -> ../bar/some_file                    # is valid
    bar/
        some_file
```

But `internal_abs_link` points to a path we *know* won't exist after install, since the source is, in nearly all cases, in the build directory and gets blown away.

Should we rewrite this?  I kind of think so, and it's simple to do and I think it would nearly always be correct.  It would be useful for packages where, for whatever reason, the build makes absolute links into the build directory.  I think this happens for packages that think they'll be built and run in the same place.

The counterargument is that none of the internal routines (`shutil` or `distutils`) actually handle this, so why should we?

Thoughts?

Reply:
Hmm... If you think it wouldn't be too much work, then there's really no reason not to do it. On the other hand, have we ever actually encountered this problem? I'm not sure if I have the time to add it myself, but if you want to add it you can.

P.S. I just realized `symlinks` defaults to True for `copy_tree` and False for `install_tree`. We should probably fix that. Which do you think is a better default?

Reply:
Personally I like `True`. Also more stuff was using `copy_tree` before, so it seems like the better choice.

Imomementing this is easy... just check whether the link target’s prefix is within the source root, and replace the source root prefix with the dest in the link target if so. It’s a couple lines.

I don’t think there is a need to handle cases where the link target is itself a link, or where it indirectly points the the source root.

I haven’t seen this come up so I’d be fine with holding off on this, though it might be worth looking at installs of packages that use install tree and seeing if any would have the problem.  Maybe save it for another PR?

Reply:
Done. Hopefully the switch from `symlinks=False` to `symlinks=True` won't break too many packages. I'm sure it will break at least a couple of packages. When it does, we can tell people to use `symlinks=False` and then we'll have example packages to test on. I'm fine with saving it for another PR.